### PR TITLE
Make array_distinct compare elements with "is distinct"

### PR DESCRIPTION
**Pull request by findepi**
Fixes https://github.com/prestodb/presto/issues/11963, fixes https://github.com/prestodb/presto/issues/11977

### CODE REVIEW

**Comment by dain** [comment]
@findepi, I'd like to see @haozhun review this.  Specifically the method handle bits, and how they relate to the new calling conventions.

Related to that, I am wondering if we can adapt the TypeSet code to alway work on an "equalitator" function based on block and position.  I think we can adapt both version to that style.

**Comment by dain** [comment]
@findepi I am no expert on these functions, but to me I would guess all of them should be using `is not distinct from` over `=`.

**Comment by haozhun** [comment]
@gerlou did the majority of work for support of elasticity around calling convention for `@OperatorDependency`. @cemcayiroglu completed a few additional pieces.

I'll take over #11263 and wrap it up. With that, `distinct from` operator will have two conventions implemented, and when annotating a parameter with `@OperatorDependency`, one can declare which one is needed.

**Comment by findepi** [comment]
@haozhun i will wait for https://github.com/prestodb/presto/pull/11983 then. Your PR is definitely bigger and harder to rebase.

@dain i will then extend this PR to cover all usages of `TypedSet` (affecting said functions).

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 1]
```diff
@@ -992,6 +992,16 @@ public void testDistinct()
         assertFunction("ARRAY_DISTINCT(ARRAY [NULL, NULL])", new ArrayType(UNKNOWN), asList((Object) null));
         assertFunction("ARRAY_DISTINCT(ARRAY [NULL, NULL, NULL])", new ArrayType(UNKNOWN), asList((Object) null));
 
+        // Indeterminate values
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Can you add the test cases mentioned in #11977  as well?

**Comment by yingsu00** [comment]
@dain @haozhun @findepi correct me if I'm wrong. I don't quite know the history of TypedSet and the isDistinctFrom operator, but it looks to me that TypedSet was trying to support the "distinctness" semantics. This can be seen from how it treats null: It has special slot in hashtable for NULL, and a flag indicating the existence of NULL, and also added guard to check NULL values before calling type.equalTo() in positionEqualsPosition() (this guard takes the distinctness semantics instead of the equal to semantics). So I would think replacing the type.equalTo with isDistinctFrom (and remove the equalTo() call) is sufficient enough, unless there is future use case that requires the "equalTo" semantics.

Another thought is: the wrong results issue (#11977) was caused by the (array.getPositionCount() == 2) optimization only. It called type.equalTo() without checking on null in front. The callsite of positionEqualsPosition() (which calls type.equalTo()) inside of TypedSet.getHashPositionOfElement() shouldn't be causing correctness issues, because it would throw "not supported" error if it sees NULL in composite data types, and for primitive types it has guard to check null in front. So if we want to fix the correctness issue before #11983 is merged, we can just add null check for that special case, or simply delete the if (array.getPositionCount() == 2) block. I can send PR for it if people think we want a quicker fix on the correctness issue.

**Comment by findepi** [comment]
>it looks to me that TypedSet was trying to support the "distinctness" semantics. 

@yingsu00 it looks so. I listed all usages above (https://github.com/prestodb/presto/pull/11979#issuecomment-441802885) and the plan is to make the class work in  "distinctness" semantics always.

> or simply delete the if (array.getPositionCount() == 2) block. I can send PR for it if people think we want a quicker fix on the correctness issue.

As you observed, this solves one known issue but not the other. (And would probably fix `array_distinct` only.)
Fortunately we have fix ready: https://github.com/prestodb/presto/pull/11983 + this. If there is a time pressure, we can defer extending the fix to other function types.

**Comment by findepi** [comment]
Superseded by https://github.com/prestosql/presto/pull/559

**Pull request closed without merging**